src/ebmc/instrument_past.cpp

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************\
+
+Module: $past Instrumentation
+
+Author: Daniel Kroening, dkr@amazon.com
+
+\*******************************************************************/
+
+#include "instrument_past.h"
+
+#include <util/arith_tools.h>
+#include <util/expr_iterator.h>
+#include <util/replace_expr.h>
+
+#include <verilog/verilog_expr.h>
+
+#include "ebmc_error.h"
+
+#include <set>
+
+// The set is ordered to get a stable numbering
+using past_sett = std::set<verilog_past_exprt>;
+
+void collect_past(const exprt &expr, past_sett &past_set)
+{
+  for(auto it = expr.depth_begin(); it != expr.depth_end(); ++it)
+  {
+    if(it->id() == ID_verilog_past)
+      past_set.insert(to_verilog_past_expr(*it));
+  }
+}
+
+void collect_past(transition_systemt &transition_system, past_sett &past_set)
+{
+  collect_past(transition_system.trans_expr, past_set);
+}
+
+void collect_past(ebmc_propertiest &properties, past_sett &past_set)
+{
+  for(auto &property : properties.properties)
+  {
+    collect_past(property.normalized_expr, past_set);
+  }
+}
+
+void complete_past_set(past_sett &past_set)
+{
+}
```

Comment:
What is this meant to do?

Reply:
Added; it generates `$past(X, i-1)` for `$past(X, i)`.